File: packages/hmssdk_flutter/ios/Classes/Actions/HMSWhiteboardAction.swift
//
//  HMSWhiteboardAction.swift
//  hmssdk_flutter
//
//  Created by Pushpam on 30/04/24.
//

import Foundation
import HMSSDK

class HMSWhiteboardAction {

    static func whiteboardActions(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {
        switch call.method {
            case "start_whiteboard":
                startWhiteboard(call, result, hmsSDK)

            case "stop_whiteboard":
                stopWhiteboard(result, hmsSDK)

            default:
                result(FlutterMethodNotImplemented)
        }
    }

    private static func startWhiteboard(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {
        let arguments = call.arguments as![AnyHashable: Any]

        /*
         title is not used as of now
         */
        guard let title = arguments["title"] as? String else {
            HMSErrorLogger.returnArgumentsError("title can't be empty")
            return
        }
        hmsSDK?.interactivityCenter.startWhiteboard {
            _, error in
            if let error = error {
                result(HMSErrorExtension.toDictionary(error))
            } else {
                result(nil)
            }
        }

    }

    private static func stopWhiteboard(_ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {

        hmsSDK?.interactivityCenter.stopWhiteboard {
            _, error in
            if let error = error {
                result(HMSErrorExtension.toDictionary(error))
            } else {
                result(nil)
            }
        }
    }

    static func getWhiteboardUpdateType(updateType: HMSWhiteboardUpdateType) -> String? {
        switch updateType {
        case .started:
            return "on_whiteboard_start"
        case .stopped:
            return "on_whiteboard_stop"
        default:
            return nil
        }
    }
}

File: packages/hmssdk_flutter/android/src/main/kotlin/live/hms/hmssdk_flutter/methods/HMSWhiteboardAction.kt
package live.hms.hmssdk_flutter.methods

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import live.hms.hmssdk_flutter.HMSCommonAction
import live.hms.hmssdk_flutter.HMSErrorLogger
import live.hms.video.sdk.HMSSDK

class HMSWhiteboardAction {
    companion object {
        fun whiteboardActions(
            call: MethodCall,
            result: Result,
            hmssdk: HMSSDK,
        ) {
            when (call.method) {
                "start_whiteboard" -> startWhiteboard(call, result, hmssdk)
                "stop_whiteboard" -> stopWhiteboard(result, hmssdk)
                else -> result.notImplemented()
            }
        }

        private fun startWhiteboard(
            call: MethodCall,
            result: Result,
            hmssdk: HMSSDK,
        ) {
            val title = call.argument<String>("title")

            title?.let {
                hmssdk.getHmsInteractivityCenter().startWhiteboard(it, HMSCommonAction.getActionListener(result))
            } ?: run {
                HMSErrorLogger.returnArgumentsError("title can't be empty")
            }
        }

        private fun stopWhiteboard(
            result: Result,
            hmssdk: HMSSDK,
        ) {
            hmssdk.getHmsInteractivityCenter().stopWhiteboard(HMSCommonAction.getActionListener(result))
        }
    }
}
